Initial import of Python Flask web app (#4)
Azure-Samples/ms-identity-docs-code-python: src/webapp/README.md

description: "This Python web application protects various routes and contacts Microsoft Graph on behalf of the user. The code in this sample is used by one or more articles on docs.microsoft.com."
products:
- azure
- azure-active-directory
- ms-graph
urlFragment: ms-identity-docs-code-webapp-python
---
-->

# Python web application - protected and accessing Microsoft Graph | Microsoft identity platform

<!-- Build badges here
![Build passing.](https://img.shields.io/badge/build-passing-brightgreen.svg) ![Code coverage.](https://img.shields.io/badge/coverage-100%25-brightgreen.svg) ![License.](https://img.shields.io/badge/license-MIT-green.svg)
-->

This sample demonstrates a Python Flask web application that is both protected by Microsoft identity platform and accesses Microsoft Graph as the user by using the Microsoft Authentication Library (MSAL) for Python..

![A browser screenshot on a page showing a response from Microsoft Graph](./app.png)

> :page_with_curl: This sample application backs one or more technical articles on docs.microsoft.com. <!-- TODO: Link to first tutorial in series when published. -->

## Prerequisites

> - Python 3.8+

Author: No Azure Subscription is required for this sample at this time.

> - Python 3.8+

Author: After reviewing your open PRs, I see you were suggesting Azure AD tenant.  I like it.  Added here.

| **Token configuration**      | _No additional claims_                                                       | This sample does not rely on any additional claims existing in the tokens.                         |
| **App roles**                | _Add an App role called `admin` for use by **Users/Groups**_                 | Required value for this sample. <bn /> One route requires your user to be assigned this role.      |

> :information_source: **Bold text** in the table matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

Use these settings in your Enterprise Application for this sample app

| Enterprise Application <br/> setting | Value for this sample app                                            | Notes                                                                                              |
|:------------------------------------:|:---------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|

> | **Users and groups**                 | _Add yourself with the **Role** of **admin**_                        | Required value for this sample. <bn /> One route requires your user to be assigned this role.      |

Reviewer: wondering whether it is now settled and we should incorporate this as well in the `csharp` samples.

Author: Not sure.  I honestly didn't look at how the csharp one was set up.  I should have cross-referenced it.  I'll leave this comment open as we get more feedback on this.

Reviewer 2: This would be the first time we direct anyone to use Enterprise Apps for adding a role and a user to that role, but that might not be a bad thing. Not sure on this, I'll need to run through it.

Author: Cool, let me know if there is a better way to do this.

Reviewer 2: Nope, this is the right way to go - ship it.

> :information_source: **Bold text** in the table matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

### 2. Update code sample with app registration values

Open the _default\_settings.py_ file and modify the three Azure Active Directory configuration properties using the values from your [app's registration in the Azure portal](https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app).

- `CLIENT_ID`: `"Application (client) ID"`
- `AUTHORITY`: `"https://login.microsoftonline.com/{your tenant}"`
- `CLIENT_CREDENTIAL`: `"********"`

### 3. Install package(s)

> To install Flask and msal into your (virtual) environment:

Reviewer: please let me quickly suggest the following change. 

```suggestion
To install Flask and MSAL libraries into your (virtual) environment:
```

Author: Good catch, will update

Reviewer: please let's mark as resolved afterwards.

Reviewer: noticed it is already resolved. marking this a solved now.

```bash
pip install -r requirements.txt
```

## Run the application

```bash
flask run --host=localhost
```

## Browse to the application

Open your browser and navigate to **http://localhost:5000**. If everything worked, the sample app should produce output similar to this:

> ![A browser screenshot showing the weclome page to the sample application.](./app.png)

Reviewer: I think it is dup.

Author: Good catch, was supposed to be home.png

description: "This Python web application protects various routes and contacts Microsoft Graph on behalf of the user. The code in this sample is used by one or more articles on docs.microsoft.com."
products:
- azure
- azure-active-directory
- ms-graph
urlFragment: ms-identity-docs-code-webapp-python
---
-->

# Python Flask | web app | user sign-in, access control (protected routes), protected web API access (Microsoft Graph) | Microsoft identity platform

<!-- Build badges here
![Build passing.](https://img.shields.io/badge/build-passing-brightgreen.svg) ![Code coverage.](https://img.shields.io/badge/coverage-100%25-brightgreen.svg) ![License.](https://img.shields.io/badge/license-MIT-green.svg)
-->

This sample demonstrates a Python Flask web application that is both protected by Microsoft identity platform and accesses Microsoft Graph as the user by using the Microsoft Authentication Library (MSAL) for Python..

![A browser screenshot on a page showing a response from Microsoft Graph](./app.png)

> :page_with_curl: This sample application backs one or more technical articles on docs.microsoft.com. <!-- TODO: Link to first tutorial in series when published. -->

## Prerequisites

> - An Azure Active Directory (Azure AD) tenant and required permissions to create App registrations.

Reviewer 2: ```suggestion
- Azure Active Directory (Azure AD) tenant and the permissions or role required for managing app registrations in the tenant.
```

|:----------------------------:|:-----------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|
| **Name**                     | `python-webapp`                                                              | Suggested value for this sample. <br/> You can change the app name at any time.                    |
| **Supported account types**  | **Accounts in this organizational directory only (Single tenant)**           | Suggested value for this sample.                                                                   |
| **Platform type**            | **Web**                                                                      | Required value for this sample. <br/> Enables the required and optional settings for the app type. |
| **Redirect URI**             | `http://localhost:5000/auth/redirect`                                        | Required value for this sample.                                                                    |

> | **Client secret**            | _Value shown in Azure portal_                                                | :warning: Record this value immediately! <br/> It's shown only _once_ (when you create it).        |

Reviewer 2: ```suggestion
| **Client secret**            | _**Value** of the client secret (not its ID)__                                                | :warning: Record this value immediately! <br/> It's shown only _once_ (when you create it).        |
```
```suggestion
| **Client secret**            | _Value shown in Azure portal_                                                | :warning: Record this value immediately! <br/> It's shown only _once_ (when you create it).        |
```

> | **Implicit grant & hybrid flows** | _None selected_                                                         | This sample does not use the implicit grant or hybrid flows.                                       |

Reviewer 2: Leave out anything they don't have to change from default.

Reviewer 2: Intent of "This sample" is unclear to me.

Author: I used that term ("this sample") from your other rows in the table where you used "this sample" to reference this thing you're looking at.  Does that help?

As you requested, I should leave this line out anyhow as _None selected_ is the default anyway.  So, I'll update this to remove this line.

Author: Updated, "default" values have been removed from this table.

| **Allow public client flows** | **No**                                                                      | This sample does not use a public client flow.                                                     |
| **Token configuration**      | _No additional claims_                                                       | This sample does not rely on any additional claims existing in the tokens.                         |

> | **App roles**                | _Add an App role called `admin` for use by **Users/Groups**_                 | Required value for this sample. <bn /> One route requires your user to be assigned this role.      |

Reviewer 2: Shorten second column to just the UI element names, if possible.

Author: Not following.  The first column is the UI element names, the second column is the value for "this sample app."  Do you have an example that you can help me see what I should change?

Reviewer 2: In short, avoid using any instructional wording in that column like the bolded ones here:

_**Add an** App role **called** `admin` **for use by** **Users/Groups**_

For app roles and possibly client secrets, we might need to use bullets in that column or even a separate table (especially if there are several roles to create). We need to be explicit for all required values, as well. For example, I also need to know what to enter for **Value** and **Description**:

![image](https://user-images.githubusercontent.com/13394541/146107774-969e6822-debd-4717-8c90-5ab6c42ee69a.png)

Author: Ah, that makes sense. Okay, tightened that slop up and added the additional fields in: https://github.com/Azure-Samples/ms-identity-docs-code-python/pull/4/commits/760ab64e273668380ed27b1b29c5051b87244d98

Reviewer 2: Dig it.

|:------------------------------------:|:---------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|
| **Users and groups**                 | _Add yourself with the **Role** of **admin**_                        | Required value for this sample. <bn /> One route requires your user to be assigned this role.      |

> :information_source: **Bold text** in the table matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

### 2. Update code sample with app registration values

Open the _default\_settings.py_ file and modify the three Azure Active Directory configuration properties using the values from your [app's registration in the Azure portal](https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app).

- `CLIENT_ID`: `"Application (client) ID"`
- `AUTHORITY`: `"https://login.microsoftonline.com/{your tenant}"`

> - `CLIENT_CREDENTIAL`: `"********"`

Reviewer 2: As noted in other comment, use single-line comments and put these in one Python-tagged code snippet.

Author: Where will this python-tagged code snippet be?  I haven't added snippets yet as we agreed that the content developer can be responsible for marking up the code so that it fits their narrative flow -- but I can put this wherever you want in anticipation.  Sorry I'm not following the suggestion for "single-line" comments and the interplay between the README (your other suggestion) and now Python-tagged code (this suggestion)

Author: Let me know if the new README + config + code layout works for you now in this regard.

### 3. Install package(s)

To install Flask and MSAL libraries into your (virtual) environment:

```bash
pip install -r requirements.txt
```

## Run the application

```bash
flask run --host=localhost
```

## Browse to the application

Open your browser and navigate to **http://localhost:5000**. If everything worked, the sample app should produce output similar to this:

![A browser screenshot showing the weclome page to the sample application.](./home.png)

## About the code

> This python web application is using the Flask web framework. It includes three navigatable routes with three different authentication/authorization levels required. As soon as the user navigates to a route that requires authentication, the user will be redirected to the Azure AD to sign in (if not already signed in via single sign-on) and accept the required app permissions.

Reviewer 2: ```suggestion
This Python web application uses the Flask web framework. The app has three routes, each requiring a different authentication/authorization level.

When a signed-out user navigates to a route requiring authentication, their browser is redirected to the Azure AD sign-in page. After signing in, and if they've not previously done so, the user is asked to consent to the app's request for permission to access their data.
```

| **Name**                       | `python-webapp`                                                              | Suggested value for this sample. <br/> You can change the app name at any time.                    |
| **Supported account types**    | **Accounts in this organizational directory only (Single tenant)**           | Suggested value for this sample.                                                                   |
| **Platform type**              | **Web**                                                                      | Required value for this sample. <br/> Enables the required and optional settings for the app type. |
| **Redirect URI**               | `http://localhost:5000/auth/redirect`                                        | Required value for this sample.                                                                    |
| **Client secret**              | _**Value** of the client secret (not its ID)_                                | :warning: Record this value immediately! <br/> It's shown only _once_ (when you create it).        |

> | **App roles**                  | **Display name**: `admin`<br/>**Allowed member types:**: **Users/Groups**<br/>**Value**: `admin`<br/>**Description**: `admin app role` | Required role for this sample. <br/> One route requires your user to be assigned this role. |

Reviewer 2: ```suggestion
| **App roles**                  | **Display name**: `admin`<br/>**Allowed member types**: **Users/Groups**<br/>**Value**: `admin`<br/>**Description**: `admin app role` | Required role for this sample. <br/> One route requires your user to be assigned this role. |
```

Yep, this format works.

| Enterprise Application <br/> setting | Value for this sample app                    | Notes                                                                                        |
|:------------------------------------:|:---------------------------------------------|:---------------------------------------------------------------------------------------------|
| **Users and groups**                 | **User**: _yourself_<br/>**Role**: **admin** | Required value for this sample. <br/> One route requires your user to be assigned this role. |

> :information_source: **Bold text** in the tables above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

### 2. Update code sample with app registration values

Open the _default\_settings.py_ file and modify the three Azure Active Directory configuration properties using the values from your [app's registration in the Azure portal](https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app).

- `CLIENT_ID`: `"Application (client) ID"`
- `CLIENT_CREDENTIAL`: `"********"`

> - `AUTHORITY`: `"https://login.microsoftonline.com/{your tenant's 'Primary domain' or 'Tenant ID'}"`

Reviewer 2: And now that you have single-line comments in the settings.py, you can include the whole block here in the README, get syntax highlighting, and it's a one-to-one match:

~~~
```Python
# 'Application (client) ID' of app registration in Azure portal - this value is a GUID
CLIENT_ID = ""

# Client secret 'Value' (not its ID) from 'Client secrets' in app registration in Azure portal
CLIENT_CREDENTIAL = ""

# Full directory URL, in the form of https://login.microsoftonline.com/<tenant>
AUTHORITY = ""
```
~~~

Renders as:

```Python
# 'Application (client) ID' of app registration in Azure portal - this value is a GUID
CLIENT_ID = ""

# Client secret 'Value' (not its ID) from 'Client secrets' in app registration in Azure portal
CLIENT_CREDENTIAL = ""

# Full directory URL, in the form of https://login.microsoftonline.com/<tenant>
AUTHORITY = ""
